ddtrace/internal/_encoding.pyx

```diff
@@ -292,175 +567,85 @@ cdef class Packer(object):
-        """Get a copy of the buffer and clear it."""
-        with self._lock:
-            try:
-                return list(self._buffer)
-            finally:
-                self._buffer[:] = []
-                self._size = 0
-
-    def encode_item(self, item): ...
-
-
-cdef class MsgpackEncoder(ListBufferedEncoder):
-    content_type = "application/msgpack"
-
-    cpdef _decode(self, data):
-        import msgpack
-        if msgpack.version[:2] < (0, 6):
-            return msgpack.unpackb(data)
-        return msgpack.unpackb(data, raw=True)
-
-    cpdef encode_trace(self, list trace):
-        return Packer().pack_trace(trace)
-
-    cpdef encode_traces(self, list traces):
-        return Packer().pack_traces(traces)
+    cpdef put(self, list trace):
+        try:
+            self._st.savepoint()
+            super(MsgpackEncoderV05, self).put(trace)
+        except Exception:
+            self._st.rollback()
+            raise
 
-    cpdef encode_item(self, item):
-        return self.encode_trace(item)
+    cdef inline int _pack_string(self, object string):
+        return msgpack_pack_uint32(&self.pk, self._st.index(string))
 
-    cpdef encode(self):
-        """Join a list of encoded objects together as a msgpack array"""
-        cdef Py_ssize_t count
-        cdef bytes buf
+    cpdef pack_span(self, object span, object dd_origin):
+        cdef int ret
 
-        with self._lock:
-            try:
-                if not self._buffer:
-                    return None
+        ret = msgpack_pack_array(&self.pk, 12)
+        if ret != 0: return ret
+
+        ret = self._pack_string(span.service)
+        if ret != 0: return ret
+        ret = self._pack_string(span.name)
+        if ret != 0: return ret
+        ret = self._pack_string(span.resource)
+        if ret != 0: return ret
+        ret = msgpack_pack_uint64(&self.pk, span.trace_id or 0)
+        if ret != 0: return ret
+        ret = msgpack_pack_uint64(&self.pk, span.span_id or 0)
+        if ret != 0: return ret
+        ret = msgpack_pack_uint64(&self.pk, span.parent_id or 0)
+        if ret != 0: return ret
+        ret = msgpack_pack_int64(&self.pk, span.start_ns or 0)
+        if ret != 0: return ret
+        ret = msgpack_pack_int64(&self.pk, span.duration_ns or 0)
+        if ret != 0: return ret
+        ret = msgpack_pack_int32(&self.pk, span.error or 0)
+        if ret != 0: return ret
+
+        ret = msgpack_pack_map(&self.pk, len(span.meta) + (dd_origin is not None))
+        if ret != 0: return ret
+        for k, v in span.meta:
+            ret = self._pack_string(k)
+            if ret != 0: return ret
+            ret = self._pack_string(v)
+            if ret != 0: return ret
+        if dd_origin is not None:
+            ret = self._pack_string(ORIGIN_KEY)
+            if ret != 0: return ret
+            ret = self._pack_string(dd_origin)
+            if ret != 0: return ret
+        
+        ret = msgpack_pack_map(&self.pk, len(span.metrics))
```

Comment:
I remember we currently don't bother encoding a dictionary for `meta` or `metrics` if there are none, is there a reason you changed it here?

Reply:
In v0.5 traces are array of fixed size 12, which means we must include empty dictionaries at the right place in the array